File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.js
// 추천인에게 배분 먼저, 나머지는 자신이 가짐
function solution(enroll, referral, seller, amount) {
    var answer = [];
    // 1. 추천인과 가입자 연결
    const relation = {};
    for (let i = 0; i < referral.length; i++) {
        relation[enroll[i]] = referral[i];
    }
    // 2. 각 인원이 가지는 금액 초기화
    const total = {};
    for (let i = 0; i < enroll.length; i++) {
        total[enroll[i]] = 0;
    }
    // 3. 배분
    for (let i = 0; i < seller.length; i++) {
        // 3-1. 각 seller의 금액(이후 재할당되므로 let 선언)
        let money = amount[i] * 100;
        // 3-2. 현재의 seller 정의
        let currSender = seller[i];
        
        // 3-3. 말단에서부터 최상단까지 배분
        while(money > 0 && currSender != "-") {
            const nextSender = relation[currSender];
            const share = Math.floor(money / 10);
            total[currSender] += money - share;
            
            money = share;
            currSender = nextSender;
        }
    }
    
    return enroll.map((name) => total[name]);
}

File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.py
def solution(enroll, referral, seller, amount):
    relation = {}
    total = {}
    
    for i in range(0, len(enroll)):
        relation[enroll[i]] = referral[i]
        total[enroll[i]] = 0
        
    for i in range(0, len(seller)):
        curr_sender = seller[i]
        money = amount[i] * 100
        
        while (money > 0 and curr_sender != "-"):
            share = money // 10
            next_sender = relation[curr_sender]
            
            total[curr_sender] += money - share
            
            money = share
            curr_sender = next_sender
        
    return [total[name] for name in enroll]